File: kranji-reading/kranji-reading-app/src/main/resources/homing/js/kranji/reading/app/read/ArticleSelectionSecretaryModule.js
// =============================================================================
// ArticleSelectionSecretaryModule — the Secretary for the article-selection
// Party.
//
// A third bus, separate from navigation and character selection. An article is
// not a tree position and not a character: the navigator's NodeSelected is a
// path, and a path means nothing to the reader, which wants an id and nothing
// else.
//
// Just a redirect: an incoming ArticleSelected is rebroadcast to every member
// as ShowArticle. Pure (state, envelope) -> Step; no DOM, no console, no
// captures.
// =============================================================================

var ArticleSelectionSecretary = {

    initial: {
        lastSelected:  null,
        recentUnknown: []
    },

    /**
     * Pure routing function. (state, envelope) -> { newState, actions }.
     */
    behavior: function (state, envelope) {
        var msg = envelope.message;

        if (msg.kind === "ArticleSelected") {
            return {
                newState: {
                    lastSelected:  msg.article,
                    recentUnknown: state.recentUnknown
                },
                actions: [{
                    kind:    "BroadcastToMembers",
                    message: { kind: "ShowArticle", article: msg.article }
                }]
            };
        }

        // Anything else is remembered for observability and otherwise ignored -
        // a bus that throws on an unknown kind couples every producer to every
        // consumer's vocabulary.
        var seen = state.recentUnknown.concat([
            { kind: msg.kind, from: envelope.from }
        ]);
        return {
            newState: {
                lastSelected:  state.lastSelected,
                recentUnknown: seen.slice(-10)
            },
            actions: []
        };
    }
};

File: kranji-reading/kranji-reading-app/src/main/java/kranji/reading/app/read/ArticleSelectionSecretaryModule.java
package kranji.reading.app.read;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the article-selection Party.
 *
 * <p>A third bus, separate from navigation and from character selection. The
 * navigator relays tree <em>positions</em>: every widget joined to it reads an
 * incoming {@code NavigateTo} as a path. An article is not a path — it is an
 * identity — and the reader wants that identity with nothing to resolve.</p>
 *
 * <p>Keeping it apart also keeps the catalogue from being the only possible
 * producer. A "continue reading" tile, a search result, or a link in another
 * article can all say {@code ArticleSelected} without knowing the tree
 * exists.</p>
 *
 * <p>Just a redirect: an incoming {@code ArticleSelected} is rebroadcast to
 * every member as {@code ShowArticle}. The producer never names its
 * consumer.</p>
 *
 * <h2>State shape</h2>
 * <pre>{@code
 * {
 *     lastSelected : "jing-ye-si" | null,
 *     recentUnknown: [{ kind, from }]     // bounded at 10
 * }
 * }</pre>
 */
public record ArticleSelectionSecretaryModule()
        implements DomModule<ArticleSelectionSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior}. */
    public record ArticleSelectionSecretary()
            implements Exportable._Constant<ArticleSelectionSecretaryModule> {}

    public static final ArticleSelectionSecretaryModule INSTANCE =
            new ArticleSelectionSecretaryModule();

    @Override
    public ImportsFor<ArticleSelectionSecretaryModule> imports() {
        return ImportsFor.<ArticleSelectionSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<ArticleSelectionSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new ArticleSelectionSecretary()));
    }
}
